Counterexamples/HomogeneousPrimeNotPrime.lean

```diff
@@ -0,0 +1,218 @@
+#align counterexample.counterexample_not_prime_but_homogeneous_prime.grading.mul_mem Counterexample.CounterexampleNotPrimeButHomogeneousPrime.grading.mul_mem
+
+end
+
+local notation "R" => ZMod 4
+
+/-- `R² ≅ {(a, a) | a ∈ R} ⨁ {(0, b) | b ∈ R}` by `(x, y) ↦ (x, x) + (0, y - x)`. -/
+def grading.decompose : R × R →+ DirectSum Two fun i => grading R i where
+  toFun zz :=
+    of (grading R ·) 0 ⟨(zz.1, zz.1), rfl⟩ +
+    of (grading R ·) 1 ⟨(0, zz.2 - zz.1), rfl⟩
+  map_zero' := by
+    refine' Dfinsupp.ext (fun (i : Two) =>
+        Option.casesOn i _ (fun (i_1 : Unit) => PUnit.casesOn i_1 _)) <;> rfl
+  map_add' := by
+    rintro ⟨a1, b1⟩ ⟨a2, b2⟩
+    rw [add_add_add_comm, ← map_add, ← map_add]
+    dsimp only [Prod.mk_add_mk]
+    simp_rw [add_sub_add_comm]
+    congr
+#align counterexample.counterexample_not_prime_but_homogeneous_prime.grading.decompose Counterexample.CounterexampleNotPrimeButHomogeneousPrime.grading.decompose
+
+theorem grading.right_inv : Function.RightInverse (coeLinearMap (grading R)) grading.decompose :=
+  fun zz => by
+  induction' zz using DirectSum.induction_on with i zz d1 d2 ih1 ih2
+  · simp only [map_zero]
+  · rcases i with (_ | ⟨⟨⟩⟩) <;> rcases zz with ⟨⟨a, b⟩, hab : _ = _⟩ <;> dsimp at hab <;>
+      cases hab <;>
+      -- Porting note: proof was `decide`
+      -- now we need a `simp` and two `erw` subproofs...
+      simp only [coeLinearMap_of, decompose, AddMonoidHom.coe_mk,
+        ZeroHom.coe_mk, sub_self, sub_zero]
+    · erw [map_zero (of (grading R ·) 1), add_zero]; rfl
+    · erw [map_zero (of (grading R ·) 0), zero_add]; rfl
+  · simp only [map_add, ih1, ih2]
+#align counterexample.counterexample_not_prime_but_homogeneous_prime.grading.right_inv Counterexample.CounterexampleNotPrimeButHomogeneousPrime.grading.right_inv
+
+theorem grading.left_inv : Function.LeftInverse (coeLinearMap (grading R)) grading.decompose :=
+  fun zz => by
+  cases' zz with a b
+  unfold grading.decompose
+  simp only [AddMonoidHom.coe_mk, ZeroHom.coe_mk, map_add, coeLinearMap_of, Prod.mk_add_mk,
+    add_zero, add_sub_cancel'_right]
+#align counterexample.counterexample_not_prime_but_homogeneous_prime.grading.left_inv Counterexample.CounterexampleNotPrimeButHomogeneousPrime.grading.left_inv
+
+instance : GradedAlgebra (grading R) where
+  one_mem := grading.one_mem R
+  mul_mem := grading.mul_mem R
+  decompose' := grading.decompose
+  left_inv := by convert grading.left_inv
+  right_inv := by convert grading.right_inv
+
+-- porting note: `I` upper case
+set_option linter.uppercaseLean3 false
+
+/-- The counterexample is the ideal `I = span {(2, 2)}`. -/
+def I : Ideal (R × R) :=
+  Ideal.span {((2, 2) : R × R)}
+#align counterexample.counterexample_not_prime_but_homogeneous_prime.I Counterexample.CounterexampleNotPrimeButHomogeneousPrime.I
+
+-- Porting note: new theorem, TODO move
+variable {G₁ : Type _} {G₂ : Type _} [CommSemigroup G₁] [CommSemigroup G₂] in
+theorem prod_dvd_iff {x y : G₁ × G₂} :
```

Comment:
I bet you don't need commutativity in this lemma. Why not move it now? What would be a good file for it?

Reply:
I have no idea what a good home would be.

Reply:
I moved it to a new file.